### PR TITLE
Fea module parameters update

### PR DESCRIPTION
**Pull request by kpwelsh**
This PR is to sort out our input parameters for the Implicit and Explicit Solvers.

The issues are:
1. We have very similar concepts defined in both, but they have slightly different implementations. They shouldn't be different.
2. We have a ton of duplicate code in all of the Simulation_Parameters* files. Primarily WRT the FEA_Modules ones.
3. We have unnecessary/unused code in here. Defining things that are unused or redundant, which makes it hard to follow.
4. Each module parses from the "root" of the input. This can lead to tricky issues with key conflicts (see next section).

### FEA Module Parameters
We had a struct called "Simulation_Parameters[_Explicit[" and one called "Simulation_Parameters_TopologyOptimization". These were two separate structures that defined simulation level parameters. The TO version inherited from the non-TO one. these things were both parsed from the root of the Yaml input file. The only crime here was that it was redundant. 

The messy-ness occurs now when we also have "FEA Module" level input parameters that _also_ inherited from Simulation_Parameters and were _also_ parsed from the root of the file. In addition to unnecessary reparsing of the same file (minor issue), this also resulted in the same Yaml input file being parsed as >2 fundamentally different data structures, which makes validation difficult and sometimes impossible.

### Materials, Regions, Boundary Conditions
Both the Implicit and Explicit solvers required users to specify:
* Boundary conditions (both Dirichlet and Neumann)
* Material properties
* Regions for material application

But they did so in different ways. In one case the material properties are applied in the module parameters. In another case, the material properties are separate in a materials definition block. In one case the boundary conditions are local to the module parameters, in another they are a single global list.

### FEA Module Parameters
Now we have completely split the module parameters and the simulation parameters. The module parameters only cover what is relevant for a specific module. Using the type discriminated input structure defined in PR #84, we can also now have individual validation for each type of FEA_Module despite it being deserialized from the same input fields. This means that
1. No more key conflicts with Simulation Parameters (i.e. unrelated structures).
2. No validation conflicts with other FEA Modules.

Now we can finally introduce the "Strict Serialization" that PR #84 promised, and give users error messages when unsupported fields are found, letting the user know that what they have input wouldn't do anything. This is very useful, as we as developers have gotten this wrong several times in the past when either placing a value somewhere incorrectly or making a typo.

### Materials, Regions, Boundary Conditions
To unify both Implicit and Explicit input parameters (where appropriate), we have merged both to a combination of the two. 
1. Material properties are specified in the materials section and aren't duplicated.
2. Regions are specified in a single location and are assigned a material.
3. FEA Modules each have their own boundary conditions.

### Optimization Options
The handling of topology optimization modules is more complete in the Implicit solver, so that structure is not used for both. This means that we can parse multi-objective modules in both solvers.

## Downstream Implementation
Not all of the possible options for inputs are supported in both solvers, although in most cases there isn't a real reason why they aren't supported and its just that they haven't been implemented yet.
In these cases, the solvers will throw an error stating that the functionality is not supported. This PR does not address the implementation of all of the parameters.

### CODE REVIEW

**Review by yencal — COMMENTED** [review]
1. I don't really get the `dynamic_options` name, I think the previous name `time_variables`  was more descriptive (but if the team is fine with dynamic_options that is fine).

2. Can we move `timer_output_level`  into the `output_options` since it is just sitting alone in the yaml file and it kinda fits in with the output_options?

3. Looking at the `Output_Options` struct
```
struct Output_Options : Yaml::DerivedFields, Yaml::ValidatedYaml {
  bool output_velocity = true;
  bool output_strain   = true;
  bool output_stress   = true;
  bool output_displacement = true;

  bool displaced_mesh      = true;

  bool output_temperature = false;
  bool output_temperature_gradient = false;
  bool output_heat_flux = false;
  
  bool strain_max       = false;
  ```
are velocity, strain, stress, displacement, temperature, temperature, temperature_gradient, heat_flux, etc not just simulation fields and should be added to `SERIALIZABLE_ENUM(SIMULATION_FIELD,`?

4. If we move all the velocity, strain, stress, displacement, temperature, temperature, temperature_gradient, heat_flux, etc into the `SERIALIZABLE_ENUM(SIMULATION_FIELD,` can we just combine `Output_Options` and `Graphics_Options`? Maybe just call it `Output_Options` since both are talking about the same thing "output". That way, the yaml file is cleaner as we would have `input_options` and `output_options` only. Or we can just put the `graphics_options` as data member of `output_options` .

5. `num_dims` is just lonely in the yaml file, can we find it a group? maybe fea_module_parameters?

6. What is the difference between `SERIALIZABLE_ENUM(FIELD,` and `SERIALIZABLE_ENUM(SIMULATION_FIELD,`? Can we have all simulation fields in one place, say we call it `SIMULATION_FIELD` then the `Ouput_Options` struct would now have a data member called `output_fields` (it kinda fits naturally, i think). Now, even if each module have different output field specific to them, since we are using a switch, a module can ignore an output field if it is not relevant to it.

7. Why was `output_fields` changed from `std::set` to `std::vector`? We don't want duplicates in the `output_fields` since some of the `ouput_fields` can require some extra calculations and it is not prudent to do the same calculation twice or more if duplicates exit in the `output_fields` variable.

8. Why was the default `output_fields` no longer available? In case the user does not specify `output_fields` in the yaml file we need to still show some default pretty pictures.

9. When I try to specify `output_fields` in the yaml file (example_simple.yaml) for example
```
output_fields:
  - velocity
  ```
I get this error
```
[cyenusah@cn307 build-fierro-serial]$ mpirun -n 1 --bind-to core ./bin/fierro-parallel-explicit example_simple.yaml 
terminate called after throwing an instance of 'Yaml::ConfigurationException'
  what():  Error in field output_fields.<0>: Provided value, `velocity` was not of allowed values: {design_density, element_id, element_switch, processor_id, speed}
  ```
This goes back to my point number 6 where i suggest we have all Solver `simulation_field` in one enum.

**Comment by kpwelsh** [comment]
@yencal 
Thanks for the feedback!

I see most of your feedback relates to the output_fields, which I agree is pretty clunky and inconsistent atm (Sometimes we have these flags, sometimes we have these sets of enums and even then we have 2 different sets of enums). Proposal for consolidation here will be at the end.

`1. I don't really get the dynamic_options name...`
I changed the name to dynamic just because I put the RK settings in there. I don't mind changing it back to time, though, since it is so similar and most of the things in there are strictly dt/time-increment related.

`2. Can we move timer_output_level into the output_options...?`
- [ ] Yes. Good suggestion.

`4. [can we combine graphics_options and output_options]`
That makes sense. I will 
- [ ] Combine graphics options with output options in a way that still separates the options relevant to the Explicit Solver from the ones relevant to ~both Explicit and Implicit.

`5. num_dims is just lonely in the yaml file, can we find it a group? maybe fea_module_parameters?`
Im pretty confident that num_dims is a simulation level parameter and not a module level parameter. If we want to change from the default we wouldn't want to have to put that in every module. Similarly, I think it belongs at a high level since it impacts both the physics simulations and the IO, so I am inclined to keep it where it is -- I don't see a problem with having a top level int.

`7. Why was output_fields changed from std::set to std::vector?`
I suspect I had a reason, but reviewing the code I see no justification + you are right about the duplicates. I will:
- [ ] Change it back.
- [ ] Add a special error message in the set deserializer to make it clear about not allowing duplicates.


### Output Fields
In the end, what I would like to see is a unification of "fields" that is integrated with both the FEA Modules, IO, and input parameters. I am drawing the line on this PR to avoid substantial changes to the FEA Modules, but that doesn't mean we can't lay the ground work here. 

In the end, I think the solution is not that complicated, and we should just have 1 list of output fields that you alluded to and consolidate the "output_*" booleans into that list. But I think we should also have a flag like "include default outputs" to let the user control whether or not the list they provide is the exclusive list or is an additive list.

So we would get rid of FEA_Module_Parameters.output_fields, and replace it with
```yaml
output_options:
  include_default_output_fields: Bool
  fields: List[FIELD]
```

Does this conflict with the changes you were working on? If so we can merge yours and then I can apply these.

**Comment by yencal** [comment]
@kpwelsh 

`5. num_dims is just lonely in the yaml file, can we find it a group? maybe fea_module_parameters?`
we can just leave it alone then since there is no natural position for it currently.

`7. Why was output_fields changed from std::set to std::vector?`
I vote to change it back because writing no need to write more error checking code.

I like this
```yaml
output_options:
  include_default_output_fields: Bool
  fields: List[FIELD]
  ```
  I think we should change `fields` to `output_fields` to matche the include_default_`output_fields`
  
  My changes was a temporary fix. I like your better. We can just polish yours and merge it in.

**Comment by Adrian-Diaz** [comment]
The changes look good to me overall. Let's make sure the default outputs remain for now. Also, the thick_condition_boundary option can be moved to optimization options; that's only relevant to TO as a constraint on the material distribution.

**Comment by Adrian-Diaz** [comment]
Also, can you replace the example_simple.yaml mesh input for mesh_Sedov_32.geo with the mesh builder equivalent to simplify this example for testing etc.?

**Comment by kpwelsh** [comment]
### Default Outputs
I added `std::set<FIELD> default_output_fields` in the FEA_Module_Parameters so each module can specify a static list of default outputs. If the input file specifies "include_default_output_fields: true" (default is true) then Simulation_Parameters will role these up into `output_options.output_fields`.

The defaults for each module are as follows:
Elasticity: displacement, displaced_mesh, strain
Heat_Conduction: temperature, heat_flux
Thermo_Elastic: displacement, displaced_mesh, strain, temperature, heat_flux
SGH: velocity, element_density, pressure, SIE, volume, mass, sound_speed
Dynamic_Elasticity: velocity, element_density, pressure, volume, mass

I think these are the defaults that would have been included 


### Mesh Building
Replaced mesh_Sedov_32.geo with
```yaml
mesh_generation_options:
  type: Box
  origin: [0, 0, 0]
  length: [1.2, 1.2, 1.2]
  num_elems: [32, 32, 32]
```

Replaced TO_small_beam.geo with 
```yaml
mesh_generation_options:
    type: Box
    origin: [0, 0, 0]
    length: [10, 10, 100]
    num_elems: [10, 10, 100]
```


### Material Properties
I went ahead an moved all of the potential material properties to the material definitions and out of the FEA_Modules. So I added these things to `material_t`:

```c++
double elastic_modulus  = 200000000000;
double poisson_ratio    = 0.3;
double density          = 7850;
double initial_temperature  = 293;
double thermal_conductivity = 10;
double specific_internal_energy_rate = 1.0;
std::vector<double> expansion_coefficients = { 12e-6, 12e-6, 12e-6, 0, 0, 0 }; 
```

The appropriate settings for the specific test problems are now included in the example*.yaml files instead of hardcoded into the FEA Module parameters.

The FEA Module Parameters now have a link to a material_t. I considered having them link to a region and then pull the material from there. That might be the way to go sometime in the future if regions mean something different, but for now it doesn't really make sense.

Lastly -- The output_fields is now a set again, and to my surprise `std::insert(item)` does not throw an error if the item is already in the set. So there doesn't need to be any checking or validation of anything here really.

**Comment by kpwelsh** [comment]
Fixed the issue where the moments of inertia werent being computed WRT the right point. 

Now implicit/explicit numbers match exactly for test problems.

**Comment by Adrian-Diaz** [comment]
I think insert to a set returns a std::pair like in the get_boundary_patches logic; it can be tested for the attempt to insert a duplicate. Other than that, is the set of output defaults additive? so having elastic and thermal together would output the combined default fields for example?

**Comment by kpwelsh** [comment]
It returns a pointer to the item in the collection. I just expected to throw like a "duplicate key" error or something. 

Output defaults is additive. 

Simulation_Parameters.h:177
```c++
  void include_default_output_fields() {
      for (auto& mod : fea_module_parameters) 
          for (auto v : mod->default_output_fields)
              output_options.output_fields.insert(v);
  }
```

We take all of the modules and add their defaults in.

If someone doesn't want some defaults, they would have to turn them all off with 
```yaml
output_options:
  include_default_output_fields: false
```

And then list all of the outputs they want.

**Review by Adrian-Diaz — APPROVED** [review]

**Review by yencal — APPROVED** [review]
